### PR TITLE
Handle upgrades with previously added charts

### PR DESCRIPTION
**Pull request by daxmc99**
If the chart has been added previously then the url of the icon will
not be included in the iconFilename. Instead of defaulting to serving
the icons, check if they have a url for the icon in the chart. If they
do, proceed to send the url to the UI and do not serve the icon data.

### CODE REVIEW

File: pkg/api/customization/catalog/template.go [hunk 1]
```diff
@@ -49,6 +112,7 @@ func TemplateFormatter(apiContext *types.APIContext, resource *types.RawResource
 
 type TemplateWrapper struct {
 	TemplateContentClient v3.TemplateContentInterface
+	TemplateVersion       v3.TemplateVersionInterface
```
Comment (cjellick) [inline pkg/api/customization/catalog/template.go]: Our convention is for the client property to be plural TemplateVersions (I realize the one right above also does not follow the convention)

File: pkg/api/customization/catalog/template.go [hunk 2]
```diff
@@ -3,20 +3,68 @@ package catalog
...
+	if i < 0 {
+		return ""
+	}
+	target = target[i:]
+	return target
+}
+
+// indexInterfaceSlice takes a list of interfaces, returns the first, asserts its a map[string]interface{}
+// and uses the provided index to retrieve the value
+func indexInterfaceSlice(value interface{}, key string) string {
```
Comment (cjellick) [inline pkg/api/customization/catalog/template.go]: looks like we could have some danger of panicking at various points in the function if you get unexpected data, do you want to add some safety checks just so we dont have to worry about that?
like, i think if you simply change `v2 := v.(map[string]interface{})` to `v2, ok := v.(map[string]interface{})`, you prevent that type conversion from panicking

File: pkg/api/server/managementstored/setup.go [hunk 3]
```diff
@@ -182,10 +182,12 @@ func Clusters(schemas *types.Schemas, managementContext *config.ScaledContext, c
 
 func Templates(schemas *types.Schemas, managementContext *config.ScaledContext) {
 	schema := schemas.Schema(&managementschema.Version, client.TemplateType)
-	schema.Formatter = catalog.TemplateFormatter
+
 	wrapper := catalog.TemplateWrapper{
 		TemplateContentClient: managementContext.Management.TemplateContents(""),
+		TemplateVersion:       managementContext.Management.TemplateVersions(""),
```
Comment (cjellick) [inline pkg/api/server/managementstored/setup.go]: lets change it to a lister. i think it should be fine

File: pkg/api/customization/catalog/template.go [hunk 4]
```diff
@@ -3,20 +3,68 @@ package catalog
...
 	"github.com/rancher/norman/types"
 	"github.com/rancher/rancher/pkg/templatecontent"
 	v3 "github.com/rancher/types/apis/management.cattle.io/v3"
 	managementschema "github.com/rancher/types/apis/management.cattle.io/v3/schema"
 	client "github.com/rancher/types/client/management/v3"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
-func TemplateFormatter(apiContext *types.APIContext, resource *types.RawResource) {
+func checkUpgrade(t TemplateWrapper, templateVersionName string) (string, error) {
```
Comment (cjellick) [inline pkg/api/customization/catalog/template.go]: checkUpgrade is a bad name with very little context in the future. Any idea for a better name?

Comment (daxmc99) [reply pkg/api/customization/catalog/template.go]: How about `getChartIconURL` ?

File: pkg/api/customization/catalog/template.go (hunk 2 above)
Comment (daxmc99) [reply pkg/api/customization/catalog/template.go]: Changed type assertion to test if the underlying type exists

File: pkg/api/customization/catalog/template.go [hunk 5]
```diff
@@ -3,35 +3,108 @@ package catalog
...
+				if err == nil {
+					resource.Links["icon"] = res
+					// also update the template itself
+					template, err := t.TemplateClients.Get(resource.ID, metav1.GetOptions{})
+					if err != nil {
+						return
+					}
+					template = template.DeepCopy()
+					template.Spec.IconFilename = res
+					_, err = t.TemplateClients.Update(template)
```
Comment (daxmc99) [inline pkg/api/customization/catalog/template.go]: We still need to update the actual `template.IconFilename` for the icon handler to behave normally. This also has an affect of charts with remote icons don't need to be parsed every time.

Comment (cjellick) [reply pkg/api/customization/catalog/template.go]: do we need to even get the err variable and return if it is not nil or can we simple completely ignore the returns of this update call?

File: pkg/api/customization/catalog/template.go [hunk 6]
```diff
@@ -3,35 +3,110 @@ package catalog
...
+					if err != nil {
+						logrus.Warnf("unable to update template icon %s", template.Name)
+						return
```
Comment (cjellick) [inline pkg/api/customization/catalog/template.go]: i dont want to return out of this function if error is not nil. just log the warning and move on

File: pkg/api/customization/catalog/template.go [hunk 7]
```diff
@@ -3,35 +3,110 @@ package catalog
...
+			ver, ok := resource.Values["versions"]
+			if ok {
+				v := indexInterfaceSlice(ver, "version")
+				res, err := getChartIconURL(t, resource.ID+"-"+v)
+				if err == nil {
+					resource.Links["icon"] = res
+					// also update the template itself
+					template, err := t.TemplateClients.Get(resource.ID, metav1.GetOptions{})
+					if err != nil {
+						logrus.Warnf("unable to get template %s", err)
```
Comment (daxmc99) [inline pkg/api/customization/catalog/template.go]: Need to use an if, else here to avoid issues if the value is nil and we try to deepcopy it

**Comment by cjellick** [comment]
lgtm

**Pull request merged**